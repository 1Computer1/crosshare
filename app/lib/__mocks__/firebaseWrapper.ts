const firebaseTesting = jest.requireActual('@firebase/testing');
import type firebaseAdminType from 'firebase-admin';

export let App: firebase.app.App;
export let AdminApp: firebaseAdminType.app.App;

export const setApp = (app: firebase.app.App) => { App = app; };
export const setAdminApp = (app: firebaseAdminType.app.App) => { AdminApp = app; };

export const AuthProvider = null;

export const DeleteSentinal = firebaseTesting.firestore.FieldValue.delete();

export const ServerTimestamp = firebaseTesting.firestore.FieldValue.serverTimestamp();

export const TimestampClass = firebaseTesting.firestore.Timestamp;
export const AdminTimestamp = firebaseTesting.firestore.Timestamp;

let anonApp: firebase.app.App;
let anonUser: firebase.User;
export const setUpForSignInAnonymously = (app: firebase.app.App, user: firebase.User) => {
  anonApp = app;
  anonUser = user;
};

export const signInAnonymously = async () => {
  setApp(anonApp);
  return anonUser;
};
