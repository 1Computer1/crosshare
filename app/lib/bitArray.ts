
const b64ToI: Array<number> = [];
const iToB64: Array<string> = [];
const numBase = '0'.charCodeAt(0);
const lowerCaseBase = 'a'.charCodeAt(0);
const upperCaseBase = 'A'.charCodeAt(0);
for (let i = 0; i <= 9; ++i) {
  b64ToI[numBase + i] = i;
  iToB64[i] = String.fromCharCode(numBase + i);
}
for (let i = 0; i < 26; ++i) {
  b64ToI[lowerCaseBase + i] = i + 10;
  iToB64[i + 10] = String.fromCharCode(lowerCaseBase + i);
  b64ToI[upperCaseBase + i] = i + 36;
  iToB64[i + 36] = String.fromCharCode(upperCaseBase + i);
}
b64ToI['-'.charCodeAt(0)] = 62;
iToB64[62] = '-';
b64ToI['_'.charCodeAt(0)] = 63;
iToB64[63] = '_';

function popCount(v: number): number {
  v -= ((v >>> 1) & 0x55555555);
  v = (v & 0x33333333) + ((v >>> 2) & 0x33333333);
  return (((v + (v >>> 4) & 0xF0F0F0F) * 0x1010101) >>> 24);
}

export class BitArray {
  constructor(private nums: Array<number>, private usedInts: number) { }

  static zero() {
    return new this([0], 1);
  }

  static fromString(input: string, base: 32 | 64) {
    if (base === 32) {
      input = input.toLowerCase();
    }
    // We only use 30 bits out of 32 max but it keeps everything simple
    const bitsPerChar = base === 32 ? 5 : 6;
    const charsPerNum = base === 32 ? 6 : 5;

    const nums: Array<number> = [];
    let usedInts = 0;

    for (let i = 0; i < input.length; ++i) {
      const index = input.length - i - 1;
      const x = b64ToI[input.charCodeAt(index)];
      const mod = i % charsPerNum;
      if (mod === 0) {
        nums[usedInts++] = x;
      } else {
        nums[usedInts - 1] |= (x << mod * bitsPerChar);
      }
    }
    return new this(nums, usedInts);
  }

  isZero() {
    return this.usedInts === 0 || (this.usedInts === 1 && this.nums[0] === 0);
  }

  clamp() {
    while (this.usedInts > 0 && this.nums[this.usedInts - 1] === 0) {
      --this.usedInts;
    }
  }

  setBit(index: number) {
    const numIndex = Math.floor(index / 30);
    const rem = index % 30;
    for (let i = this.usedInts; i < numIndex + 1; i += 1) {
      this.usedInts += 1;
      this.nums.push(0);
    }
    this.nums[numIndex] |= (1 << rem);
  }

  toString(base: 32 | 64) {
    const bitsPerChar = base === 32 ? 5 : 6;
    const charsPerNum = base === 32 ? 6 : 5;
    const mask = (1 << bitsPerChar) - 1;
    let s = '';
    for (let i = this.usedInts - 1; i >= 0; --i) {
      for (let j = charsPerNum - 1; j >= 0; --j) {
        const char = iToB64[(this.nums[i] >> (j * bitsPerChar)) & mask];
        if (s || char !== '0') {
          s += char;
        }
      }
    }
    return s || '0';
  }

  bitLength(): number {
    if (!this.usedInts) {
      return 0;
    }
    return 30 * (this.usedInts - 1) + (32 - Math.clz32(this.nums[this.usedInts - 1]));
  }

  /** TODO we can make a faster version if we expect bit array to be sparse */
  bitCount(): number {
    let sum = 0;
    for (let i = 0; i < this.usedInts; i += 1) {
      if (this.nums[i]) {
        sum += popCount(this.nums[i]);
      }
    }
    return sum;
  }

  and(other: BitArray) {
    const nums: Array<number> = [];
    const usedInts = Math.min(this.usedInts, other.usedInts);

    for (let i = 0; i < usedInts; ++i) {
      nums[i] = this.nums[i] & other.nums[i];
    }

    const res = new BitArray(nums, usedInts);
    res.clamp();
    return res;
  }

  activeBits(): Array<number> {
    const ret = [];
    for (let i = this.usedInts - 1; i >= 0; i--) {
      let num = this.nums[i];
      while (num !== 0) {
        const t = 31 - Math.clz32(num);
        num ^= 1 << t;
        ret.push((i * 30) + t);
      }
    }
    return ret;
  }
}
