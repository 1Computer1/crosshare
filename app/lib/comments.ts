import {
  CommentForModerationWithIdT,
  CommentWithRepliesT,
  DBPuzzleT,
  DBPuzzleV,
} from './dbtypes';
import { getFromDB } from './dbUtils';

function findCommentById(
  comments: Array<CommentWithRepliesT>,
  id: string
): CommentWithRepliesT | null {
  for (const comment of comments) {
    if (comment.i === id) {
      return comment;
    }
    if (comment.r !== undefined) {
      const res = findCommentById(comment.r, id);
      if (res !== null) {
        return res;
      }
    }
  }
  return null;
}

export async function moderateComments(
  db: FirebaseFirestore.Firestore,
  commentsForModeration: Array<CommentForModerationWithIdT>,
  commentIdsForDeletion: Set<string>,
  autoModerating: boolean
) {
  const puzzles: Record<string, DBPuzzleT> = {};
  if (commentsForModeration) {
    for (const comment of commentsForModeration) {
      // Don't need to do anything for any comment that has been marked for deletion
      if (!commentIdsForDeletion.has(comment.i)) {
        let puzzle: DBPuzzleT | null = null;
        const fromCache = puzzles[comment.pid];
        if (fromCache) {
          puzzle = fromCache;
        } else {
          try {
            puzzle = await getFromDB('c', comment.pid, DBPuzzleV);
            puzzles[comment.pid] = puzzle;
          } catch {
            puzzle = null;
          }
        }
        if (puzzle) {
          if (puzzle.cs === undefined) {
            puzzle.cs = [];
          }
          if (comment.rt === null) {
            puzzle.cs.push(comment);
          } else {
            const parent = findCommentById(puzzle.cs, comment.rt);
            if (parent === null) {
              throw new Error('parent comment not found');
            }
            if (parent.r) {
              parent.r.push(comment);
            } else {
              parent.r = [comment];
            }
          }
        }
      }
      await db.collection('cfm').doc(comment.i).delete();
      if (autoModerating) {
        db.collection('automoderated').doc(comment.i).create(comment);
      }
    }
  }

  // Now we've merged in all the comments, so update the puzzles:
  for (const [puzzleId, dbPuzzle] of Object.entries(puzzles)) {
    console.log('updating', puzzleId, dbPuzzle.cs);
    await db.collection('c').doc(puzzleId).update({ cs: dbPuzzle.cs });
  }
}
