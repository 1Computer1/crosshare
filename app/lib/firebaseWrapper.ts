declare var firebase: typeof import('firebase');

// On the server we need to require firebase - on client it's done w/ script tags
if (typeof window === "undefined") {
  globalThis.firebase = require("firebase");
  require("firebase/firestore");
}

const firebaseConfig = {
  apiKey: "AIzaSyBrmmBf91peVT5T_Z7N3z9oizsPH5u2pUc",
  authDomain: "auth.crosshare.org",
  databaseURL: "https://mdcrosshare.firebaseio.com",
  projectId: "mdcrosshare",
  storageBucket: "mdcrosshare.appspot.com",
  messagingSenderId: "603173482014",
  appId: "1:603173482014:web:98d7d820731b7c5eaa080f",
  measurementId: "G-LTLN7Z4XBS"
};

// Initialize Firebase
export let App: firebase.app.App;
if (firebase.apps.length) {
  App = firebase.apps[0];
} else {
  App = firebase.initializeApp(firebaseConfig);
  if (typeof window !== "undefined") {
    if (process.env.NODE_ENV === 'production') {
      App.performance();
    }
  }
  if (process.env.NEXT_PUBLIC_USE_FIREBASE_EMULATOR) {
    var db = App.firestore();
    db.settings({
      host: "localhost:8080",
      ssl: false
    });
  }
}

export const setApp = (app: firebase.app.App) => { App = app };

export const AuthProvider = new firebase.auth.GoogleAuthProvider();

export const DeleteSentinal = firebase.firestore.FieldValue.delete();

export const TimestampClass = firebase.firestore.Timestamp;
export type TimestampType = firebase.firestore.Timestamp;
