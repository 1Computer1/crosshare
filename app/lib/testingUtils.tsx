import { ReactElement, ComponentType } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { matchers, createSerializer } from 'jest-emotion';

import { AuthContext } from '../components/AuthContext';

import type firebaseTypes from 'firebase';

// Add the custom matchers provided by 'jest-emotion'
expect.extend(matchers);
expect.addSnapshotSerializer(createSerializer());

export const anonymousUser = {
  uid: 'anonymous-user-id',
  isAnonymous: true
} as firebaseTypes.User;

const WithAllProviders: ComponentType = ({ children }) => {
  return (
    <AuthContext.Provider value={{ user: anonymousUser, isAdmin: false, loadingUser: false, error: undefined }}>
      {children}
    </AuthContext.Provider>
  )
}

function wrappedRender(
  ui: ReactElement,
  options?: RenderOptions
): RenderResult {

  return render(ui, {
    wrapper: WithAllProviders,
    ...options,
  });
}
// re-export everything
export * from '@testing-library/react';

// override render method
export { wrappedRender as render };
