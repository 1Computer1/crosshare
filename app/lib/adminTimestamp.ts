/* this is the firebase timestamp type for the main react project! */

import * as t from 'io-ts';
import { either } from 'fp-ts/lib/Either';

import { AdminTimestamp } from './firebaseWrapper';
import type firebaseAdminType from 'firebase-admin';

const isFirestoreTimestamp = (u: unknown): u is firebaseAdminType.firestore.Timestamp =>
  u ? u instanceof AdminTimestamp : false;

const validateTimestamp: t.Validate<unknown, firebaseAdminType.firestore.Timestamp> = (i, c) => {
  if (isFirestoreTimestamp(i)) {
    return t.success(i);
  }
  return either.chain(
    t.type({ seconds: t.number, nanoseconds: t.number }).validate(i, c),
    obj => t.success(new AdminTimestamp(obj.seconds, obj.nanoseconds))
  );
};

export const adminTimestamp = new t.Type<firebaseAdminType.firestore.Timestamp>(
  'AdminTimestamp',
  isFirestoreTimestamp,
  validateTimestamp,
  t.identity,
);
