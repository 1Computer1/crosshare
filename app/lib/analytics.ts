/* This gets used by the analytics cron function in the `functions` directory.

It lives here so we can test it. */
import type firebase from 'firebase-admin';
import { PathReporter } from 'io-ts/lib/PathReporter';
import { isRight } from 'fp-ts/lib/Either';

import {
  LegacyPlayV,
  PuzzleStatsT,
  PuzzleStatsV,
  DailyStatsV,
  DailyStatsT,
  DBPuzzleV,
  DBPuzzleT,
  getDateString,
} from './dbtypes';

import { TimestampType } from './firebaseWrapper';

export async function runAnalytics(
  db: firebase.firestore.Firestore,
  startTimestamp: TimestampType,
  endTimestamp: TimestampType
) {
  console.log(
    'Updating analytics btwn',
    startTimestamp.toDate().toLocaleString(),
    endTimestamp.toDate().toLocaleString()
  );
  const value = await db
    .collection('p')
    .where('f', '==', true)
    .where('ua', '>=', startTimestamp)
    .where('ua', '<', endTimestamp)
    .orderBy('ua', 'asc')
    .get();
  console.log('Updating analytics for ' + value.size + ' plays');

  const puzzleMap: Map<string, DBPuzzleT> = new Map();
  const puzzleStatsMap: Map<string, PuzzleStatsT> = new Map();
  const dailyStatsMap: Map<string, DailyStatsT> = new Map();

  // Get puzzle obj from cache or db
  async function getPuzzle(puzzleId: string): Promise<DBPuzzleT> {
    const puzzle = puzzleMap.get(puzzleId);
    if (puzzle) {
      return puzzle;
    }
    const puzzleRes = await db.collection('c').doc(puzzleId).get();
    if (!puzzleRes.exists) {
      throw new Error('Missing puzzle but have play: ' + puzzleId);
    }
    const dbpuzzle = DBPuzzleV.decode(puzzleRes.data());
    if (!isRight(dbpuzzle)) {
      console.error(PathReporter.report(dbpuzzle).join(','));
      throw new Error('Malformed puzzle');
    }
    puzzleMap.set(puzzleId, dbpuzzle.right);
    return dbpuzzle.right;
  }

  for (const doc of value.docs) {
    const validationResult = LegacyPlayV.decode(doc.data());
    if (!isRight(validationResult)) {
      console.error(PathReporter.report(validationResult).join(','));
      throw new Error('Malformed play');
    }
    const play = validationResult.right;

    // Make sure we have a puzzle stats obj in the cache for the relevant puzzle
    let puzzleStats = puzzleStatsMap.get(play.c);
    if (!puzzleStats) {
      // get puzzle stats from db or create a new stats object
      const psvalue = await db.collection('s').doc(play.c).get();
      if (psvalue.exists) {
        const result = PuzzleStatsV.decode(psvalue.data());
        if (!isRight(result)) {
          console.error(PathReporter.report(result).join(','));
          throw new Error('Malformed puzzle stats');
        }
        puzzleStats = result.right;
        puzzleStatsMap.set(play.c, puzzleStats);
      } else {
        const puzzle = await getPuzzle(play.c);
        puzzleStats = {
          a: puzzle.a,
          ua: endTimestamp,
          n: 0,
          s: 0,
          nt: 0,
          st: 0,
          ct: [],
          uc: [],
          sct: Array.from(
            { length: 6 },
            () => Math.random().toString(36)[2]
          ).join(''),
        };
        puzzleStatsMap.set(play.c, puzzleStats);
      }
    }

    // don't count the author's play in puzzle stats
    if (puzzleStats.a === play.u) {
      continue;
    }

    puzzleStats.ua = endTimestamp;
    puzzleStats.n += 1;
    puzzleStats.nt += play.t;
    if (!play.ch) {
      puzzleStats.s += 1;
      puzzleStats.st += play.t;
    }
    const maxTime = Math.max(...play.ct);
    const maxCount = Math.max(...play.uc);
    for (let i = 0; i < play.ct.length; i += 1) {
      let updateTime = play.ct[i];
      const updateIters = play.uc[i];
      if (play.rc.indexOf(i) !== -1 || play.we.indexOf(i) !== -1) {
        /* If a cell was revealed or checked & wrong, make it's update time the
         * end of the play. This way cheat cells always show as taking the
         * longest for the user in question. */
        updateTime = play.t;
      }
      puzzleStats.ct[i] = (puzzleStats.ct[i] || 0) + updateTime / maxTime;
      puzzleStats.uc[i] = (puzzleStats.uc[i] || 0) + updateIters / maxCount;
    }

    // Next update daily stats for the relevant date
    const pd = play.ua.toDate();
    const dateString = getDateString(pd);
    let dailyStats = dailyStatsMap.get(dateString);
    if (!dailyStats) {
      // get daily stats from db or create a new stats object
      const dsvalue = await db.collection('ds').doc(dateString).get();
      if (dsvalue.exists) {
        const result = DailyStatsV.decode(dsvalue.data());
        if (!isRight(result)) {
          console.error(PathReporter.report(result).join(','));
          throw new Error('Malformed daily stats');
        }
        dailyStats = result.right;
        dailyStatsMap.set(dateString, dailyStats);
      } else {
        dailyStats = {
          ua: endTimestamp,
          n: 0,
          u: [],
          c: {},
          i: {},
          h: Array<number>(24).fill(0),
        };
        dailyStatsMap.set(dateString, dailyStats);
      }
    }
    dailyStats.ua = endTimestamp;
    dailyStats.n += 1;
    if (dailyStats.u.indexOf(play.u) === -1) {
      dailyStats.u.push(play.u);
    }
    dailyStats.c[play.c] = (dailyStats.c[play.c] || 0) + 1;
    if (dailyStats.i && !dailyStats.i[play.c]) {
      const puzzle = await getPuzzle(play.c);
      dailyStats.i[play.c] = [puzzle.t, puzzle.n, puzzle.a];
    }
    const hour = pd.getUTCHours();
    dailyStats.h[hour] = (dailyStats.h[hour] || 0) + 1;
  }

  console.log('Done, writing ' + puzzleStatsMap.size + ' puzzle stats objects');
  for (const [crosswordId, puzzleStats] of puzzleStatsMap.entries()) {
    await db.collection('s').doc(crosswordId).set(puzzleStats);
  }
  console.log('Writing ' + dailyStatsMap.size + ' daily stats objects');
  for (const [dateString, dailyStats] of dailyStatsMap.entries()) {
    await db.collection('ds').doc(dateString).set(dailyStats);
  }
}
