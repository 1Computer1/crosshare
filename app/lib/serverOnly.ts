import { AdminApp, AdminTimestamp } from '../lib/firebaseWrapper';
import { PuzzleResult, puzzleFromDB } from './types';
import type firebaseAdminType from 'firebase-admin';

import * as t from 'io-ts';
import { isRight } from 'fp-ts/lib/Either';
import { PathReporter } from 'io-ts/lib/PathReporter';
import { DBPuzzleV, DBPuzzleT } from './dbtypes';
import { adminTimestamp } from './adminTimestamp';
import { mapEachResult } from './dbUtils';
import { ConstructorPageT, ConstructorPageV } from './constructorPage';

export async function getStorageUrl(storageKey: string): Promise<string | null> {
  const profilePic = AdminApp.storage().bucket().file(storageKey);
  if ((await profilePic.exists())[0]) {
    try {
      return (await profilePic.getSignedUrl({
        action: 'read',
        expires: '03-09-2491'
      }))[0];
    } catch (e) {
      console.log('error getting profile pic', storageKey, e);
    }
  } else {
    console.log('pic doesnt exist', storageKey);
  }
  return null;
}

const PuzzleIndexV = t.type({
  /** array of puzzle timestamps */
  t: t.array(adminTimestamp),
  /** array of puzzle ids */
  i: t.array(t.string)
});
type PuzzleIndexT = t.TypeOf<typeof PuzzleIndexV>;

async function getPuzzlesForPage(indexDocId: string, queryField: string, queryValue: string | boolean, page: number, page_size: number): Promise<[Array<PuzzleResult>, PuzzleIndexT]> {
  const db = AdminApp.firestore();
  const indexDoc = await db.collection('i').doc(indexDocId).get();
  let index: PuzzleIndexT | null = null;
  if (indexDoc.exists) {
    const validationResult = PuzzleIndexV.decode(indexDoc.data());
    if (isRight(validationResult)) {
      index = validationResult.right;
    } else {
      console.error(PathReporter.report(validationResult).join(','));
      throw new Error('failed to validate index for ' + indexDocId);
    }
  }
  if (index === null) {
    console.log('No index, initializing', indexDocId);
    index = { t: [], i: [] };
  }

  let q = db.collection('c').where(queryField, '==', queryValue).orderBy('p', 'desc');
  if (index.i.length) {
    const mostRecentTimestamp = index.t[0];
    if (mostRecentTimestamp) {
      q = q.endBefore(mostRecentTimestamp);
    }
  }
  const newPuzzles: Array<DBPuzzleT & { id: string }> = await mapEachResult(q, DBPuzzleV, (dbpuzz, docId) => {
    return { ...dbpuzz, id: docId };
  });

  if (newPuzzles.length) {
    console.log(`Adding ${newPuzzles.length} to index for ${indexDocId}`);
    // Add new puzzles to the beginning
    for (const p of newPuzzles.reverse()) {
      index.t.unshift(AdminTimestamp.fromMillis(p.p.toMillis()));
      index.i.unshift(p.id);
    }
    await db.collection('i').doc(indexDocId).set(index);
  }
  const start = page * page_size;
  const entriesForPage = index.i.slice(start, start + page_size);

  const puzzles: Array<PuzzleResult> = [];

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  for (const puzzleId of entriesForPage) {
    const alreadyHave = newPuzzles.find(x => x.id === puzzleId);
    if (alreadyHave) {
      puzzles.push({ ...puzzleFromDB(alreadyHave), id: alreadyHave.id });
      continue;
    }

    const dbres = await db.collection('c').doc(puzzleId).get();
    if (!dbres.exists) {
      console.warn('Puzzle id in index but no puzzle exists ', puzzleId);
      continue;
    }
    const validationResult = DBPuzzleV.decode(dbres.data());
    if (isRight(validationResult)) {
      puzzles.push({ ...puzzleFromDB(validationResult.right), id: dbres.id });
    } else {
      console.error('Puzzle id in index but invalid puzzle', puzzleId);
      console.error(PathReporter.report(validationResult).join(','));
    }
  }
  return [puzzles, index];
}

export async function getPuzzlesForConstructorPage(userId: string, page: number, page_size: number): Promise<[Array<PuzzleResult>, PuzzleIndexT]> {
  return getPuzzlesForPage(userId, 'a', userId, page, page_size);
}

export async function getPuzzlesForFeatured(page: number, page_size: number): Promise<[Array<PuzzleResult>, PuzzleIndexT]> {
  return getPuzzlesForPage('featured', 'f', true, page, page_size);
}

const usernameMap: Record<string, ConstructorPageT> = {};
let usernamesUpdated: number | null = null;
const usernamesTTL = 1000 * 60 * 10;

export async function userIdToPage(userId: string): Promise<ConstructorPageT | null> {
  const now = Date.now();
  if (usernamesUpdated === null || now - usernamesUpdated > usernamesTTL) {
    const db = AdminApp.firestore();
    let query: firebaseAdminType.firestore.Query = db.collection('cp');
    if (usernamesUpdated) {
      query = query.where('t', '>=', AdminTimestamp.fromMillis(usernamesUpdated));
    }
    try {
      await mapEachResult(query,
        ConstructorPageV, (cp, docId) => {
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const { t, ...partial } = cp;
          usernameMap[cp.u] = { ...partial, id: docId };
        });
      usernamesUpdated = now;
    } catch (e) {
      console.error('error updating constructor pages');
      console.error(e);
    }
  }
  return usernameMap[userId] || null;
}

export async function sendEmail(toAddress: string, subject: string, text: string, html: string) {
  const db = AdminApp.firestore();
  return db.collection('mail').add({
    to: [toAddress],
    message: { subject, text, html }
  });
}
