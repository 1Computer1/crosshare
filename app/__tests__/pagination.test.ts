import * as firebaseTesting from '@firebase/testing';
import type firebaseAdminType from 'firebase-admin';

import { setAdminApp } from '../lib/firebaseWrapper';
//import { getUser, render, cleanup, fireEvent } from '../lib/testingUtils';

const projectId = 'paginationtests';

jest.mock('../lib/firebaseWrapper');

import { getPuzzlesForFeatured } from '../lib/serverOnly';
import { getMockedPuzzle } from '../lib/testingUtils';

test('try it when empty', async () => {
  await firebaseTesting.clearFirestoreData({ projectId });
  const adminApp = firebaseTesting.initializeAdminApp({ projectId }) as firebase.app.App;
  setAdminApp(adminApp as unknown as firebaseAdminType.app.App);

  expect((await adminApp.firestore().collection('i').doc('featured').get()).exists).toBeFalsy();

  await getPuzzlesForFeatured(0, 2);

  expect((await adminApp.firestore().collection('i').doc('featured').get()).exists).toBeFalsy();

  adminApp.delete();
});

test('try it with a few puzzles', async () => {
  await firebaseTesting.clearFirestoreData({ projectId });
  const adminApp = firebaseTesting.initializeAdminApp({ projectId }) as firebase.app.App;
  setAdminApp(adminApp as unknown as firebaseAdminType.app.App);

  expect((await adminApp.firestore().collection('i').doc('featured').get()).exists).toBeFalsy();

  await adminApp.firestore().collection('c').doc('featured1').set(getMockedPuzzle({ f: true }));
  await adminApp.firestore().collection('c').doc('nonfeatured').set(getMockedPuzzle());
  await adminApp.firestore().collection('c').doc('featured2').set(getMockedPuzzle({ f: true }));
  await adminApp.firestore().collection('c').doc('featured3').set(getMockedPuzzle({ f: true }));

  let [page1] = await getPuzzlesForFeatured(0, 2);
  expect(page1.map(p => p.id)).toEqual(['featured3', 'featured2']);
  let [page2] = await getPuzzlesForFeatured(1, 2);
  expect(page2.map(p => p.id)).toEqual(['featured1']);

  expect((await adminApp.firestore().collection('i').doc('featured').get()).data() ?.i).toMatchSnapshot();

  await adminApp.firestore().collection('c').doc('featured4').set(getMockedPuzzle({ f: true }));
  await adminApp.firestore().collection('c').doc('nonfeatured2').set(getMockedPuzzle());
  await adminApp.firestore().collection('c').doc('featured5').set(getMockedPuzzle({ f: true }));

  [page1] = await getPuzzlesForFeatured(0, 2);
  expect(page1.map(p => p.id)).toEqual(['featured5', 'featured4']);
  [page2] = await getPuzzlesForFeatured(1, 2);
  expect(page2.map(p => p.id)).toEqual(['featured3', 'featured2']);

  expect((await adminApp.firestore().collection('i').doc('featured').get()).data() ?.i).toMatchSnapshot();

  adminApp.delete();
});
