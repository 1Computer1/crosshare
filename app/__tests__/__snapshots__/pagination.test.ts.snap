// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`try it with a few puzzles 1`] = `
Array [
  "featured3",
  "featured2",
  "featured1",
]
`;

exports[`try it with a few puzzles 2`] = `
Array [
  "featured5",
  "featured4",
  "featured3",
  "featured2",
  "featured1",
]
`;
